components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt

```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
         }
     }
 
-    private val guid by lazy { safeArguments.getString(KEY_LOGIN_GUID) }
+    private val dispatcher = dispatcher
     private val origin by lazy { safeArguments.getString(KEY_LOGIN_ORIGIN)!! }
     private val formActionOrigin by lazy { safeArguments.getString(KEY_LOGIN_FORM_ACTION_ORIGIN) }
     private val httpRealm by lazy { safeArguments.getString(KEY_LOGIN_HTTP_REALM) }
     @VisibleForTesting
     internal val icon by lazy { safeArguments.getParcelable<Bitmap>(KEY_LOGIN_ICON) }
-
     @VisibleForTesting
     internal var username by SafeArgString(KEY_LOGIN_USERNAME)
     @VisibleForTesting
     internal var password by SafeArgString(KEY_LOGIN_PASSWORD)
-
     private var validateStateUpdate: Job? = null
+    // List of logins already stored for this site.  We use this to determine
+    // how to save the new [LoginEntry] (add a new record, or update an existing record)
+    private var loginsForOrigin: Deferred<List<Login>>? = null
+
+    // Start the process to fetch `loginsForOrigin` if we need to.
+    // This makes sure that:
+    //   - We don't start the process before `feature.loginsStorage` is set
+    //   - We only run the process once
+    private fun fetchLoginsForOriginIfNeeded() {
+        val loginsStorage = feature?.loginsStorage
+        if (loginsStorage != null && loginsForOrigin == null) {
+            val loginsForOrigin = CoroutineScope(dispatcher).async {
+                loginsStorage.decryptLogins(loginsStorage.getByBaseDomain(origin))
+            }
+            loginsForOrigin.invokeOnCompletion {
+                // Make sure to run update() now that the logins have been fetched
+                CoroutineScope(Main).launch {
+                    update()
```

Comment:
My first instinct is to move this elsewhere and restructure this code a bit. I.e. separate stuff that updates UI from stuff that manages and mutates local state.

As-is, you now have a side-effect to this function that changes UI state, which is a potential footgun. E.g. it's not clear now at what points UI will actually change - simply calling `findLogin` method can end up invoking this, which is _maybe_ what you want, but seems a bit surprising to me.

Reply:
I like that idea a lot and it would also make the testing easier.  I was afraid to do that when I was coding that, mostly because I'm not sure how the Android lifecycle works.  If I split out a class that handles the dupe-checking, when would I construct it? It would need to be after the feature is set but before the view needs updating.  Would `onViewCreated` work?  Is that called when Android is recreating the fragment?